README.md
**Pyortho**
======

## Description

**Pyortho** is a python package for local signal-and-noise orthogonalization and local similarity calculation. The local orthogonalization method is a fundamental seismic data analysis algorithm and has a wide range of applications. More examples will be continuously updated. 

## Reference
    Chen, Y. and S. Fomel. "Random noise attenuation using local signal-and-noise orthogonalization." Geophysics 80, no. 6 (2015): WD1-WD9.

BibTeX:

	@article{chen2015random,
	  title={Random noise attenuation using local signal-and-noise orthogonalization},
	  author={Chen, Yangkang and Fomel, Sergey},
	  journal={Geophysics},
	  volume={80},
	  number={6},
	  pages={WD1--WD9},
	  year={2015},
	  publisher={Society of Exploration Geophysicists}
	}

-----------
## Copyright
    Initial version: Yangkang Chen (chenyk2016@gmail.com), 2021-2022
	Later version: pyortho developing team, 2022-present
-----------

## License
    GNU General Public License, Version 3
    (http://www.gnu.org/copyleft/gpl.html)   

-----------

## Install
Using the latest version

    git clone https://github.com/chenyk1990/pyortho
    cd pyortho
    pip install -v -e .
or using Pypi

    pip install pyortho

-----------
## Examples
    The "demo" directory contains all runable scripts to demonstrate different applications of pyortho. 

-----------
## Dependence Packages
* scipy 
* numpy 
* matplotlib

-----------
## Development
    The development team welcomes voluntary contributions from any open-source enthusiast. 
    If you want to make contribution to this project, feel free to contact the development team. 

-----------
## Contact
    Regarding any questions, bugs, developments, collaborations, please contact  
    Yangkang Chen
    chenyk2016@gmail.com

-----------




CHANGELOG.md
Changelog
---------

* **Pyortho 0.0.3** *Apr 16 2022*
    * Some bugfixes. 

* **Pyortho 0.0.1** *Mar 16 2022*
    * Initial Pyseistr release.


test_localortho.py
#  DEMO script (python version) for calculating local orthogonalization and its application in random noise attenuation 
#  
#  Copyright (C) 2022 Yangkang Chen
#  
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published
#  by the Free Software Foundation, either version 3 of the License, or
#  any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details: http://www.gnu.org/licenses/
#  
#  Reference:   1. Random noise attenuation using local signal-and-noise orthogonalization
#               Chen and Fomel, 2015, Geophysics
#               2. Ground-Roll Noise Attenuation Using a Simple and Effective Approach Based on 
#               Local Band-Limited Orthogonalization, Chen et al., 2015, IEEE Geoscience and Remote Sensing Letters
#               3. Iterative deblending with multiple constraints based on shaping regularization,
#               Chen, 2015, IEEE Geoscience and Remote Sensing Letters
#               4. Orthogonalized morphological reconstruction for weak signal detection in micro-seismic monitoring:
#               Methodology, Huang et al., 2018, GJI
#               5. Surface-related multiple leakage extraction using local primary-and-multiple 
#               orthogonalization, Zhang et al., 2020, Geophysics
#               6. Non-stationary local signal-and-noise orthogonalization, Chen et al.,
#               2021, Geophysics
#               7. Local primary-and-multiple orthogonalization for leaked internal multiple crosstalk estimation and attenuation on full-wavefield migrated images
#               Zhang, et al., 2021, Geophysics

## generate synthetic data
#This synthetic data was used in Huang et al., 2016, Damped multichannel singular spectrum analysis for 3D random noise attenuation, Geophysics, 81, V261-V270.
import numpy as np
import matplotlib.pyplot as plt


## generate the synthetic data
a1=np.zeros([300,80])
[n,m]=a1.shape
a3=a1.copy();
a4=a1.copy();

k=-1;
a=0.1;
b=1;
pi=np.pi

ts=np.arange(-0.055,0.055+0.002,0.002)
b1=np.zeros([len(ts)])
b2=np.zeros([len(ts)])
b3=np.zeros([len(ts)])
b4=np.zeros([len(ts)])

for t in ts:
    k=k+1;
    b1[k]=(1-2*(pi*30*t)*(pi*30*t))*np.exp(-(pi*30*t)*(pi*30*t));
    b2[k]=(1-2*(pi*40*t)*(pi*40*t))*np.exp(-(pi*40*t)*(pi*40*t));
    b3[k]=(1-2*(pi*40*t)*(pi*40*t))*np.exp(-(pi*40*t)*(pi*40*t));
    b4[k]=(1-2*(pi*30*t)*(pi*30*t))*np.exp(-(pi*30*t)*(pi*30*t));

t1=np.zeros([m],dtype='int')
t3=np.zeros([m],dtype='int')
t4=np.zeros([m],dtype='int')
for i in range(m):
  t1[i]=np.round(140);
  t3[i]=np.round(-2*i+220);
  t4[i]=np.round(2*i+10);
  a1[t1[i]:t1[i]+k+1,i]=b1; 
  a3[t3[i]:t3[i]+k+1,i]=b1; 
  a4[t4[i]:t4[i]+k+1,i]=b1; 

d0=a1[0:300,:]+a3[0:300,:]+a4[0:300,:];

## add noise
[n1,n2]=d0.shape
np.random.seed(201415)
n=0.1*np.random.randn(n1,n2);
dn=d0+n;
print(np.std(dn))


from fxydmssa import fxydmssa
d1=fxydmssa(dn,0,120,0.004,3,1);	#DMSSA (when damping factor =1, there are heavy damages)
noi1=dn-d1;


## calculate local orthogonalization
from localortho import localortho
rect=[20,20,1];
eps=0;
niter=20;
verb=1;
[d2,noi2,low]=localortho(d1,noi1,rect,niter,eps,verb);
## Use Python 3 (Windows)
d2 = np.squeeze(d2)  
noi2 = np.squeeze(noi2)
## calculate local similarity
from localsimi import localsimi
simi1=localsimi(d1,noi1,[5,5,1],niter,eps,verb);
simi2=localsimi(d2,noi2,[5,5,1],niter,eps,verb);
## Use Python 3 (Windows)
simi1 = np.squeeze(simi1)  
simi2 = np.squeeze(simi2)
## compare SNR
from str_snr import str_snr
print('SNR of initial denoising is %g'%str_snr(d0,d1));
print('SNR of local orthogonalization is %g'%str_snr(d0,d2));

## compare with matlab
import scipy
from scipy import io
datas = {"d0":d0,"dn": dn, "d1": d1, "noi1": noi1, "d2":d2, "noi2":noi2}
scipy.io.savemat("datas2d.mat", datas)


## plot results
fig = plt.figure(figsize=(10, 5))
ax = fig.add_subplot(3,2,1)
# ax.set_xticks([])
# ax.set_yticks([])
plt.imshow(dn,cmap='jet',clim=(-0.2, 0.2),aspect=0.2)
plt.title('Noisy data');
fig.add_subplot(3,2,3)
plt.imshow(d1,cmap='jet',clim=(-0.2, 0.2),aspect=0.2)
plt.title('Initial denoising');
fig.add_subplot(3,2,4)
plt.imshow(noi1,cmap='jet',clim=(-0.2, 0.2),aspect=0.2)
plt.title('Initial denoising');
plt.imshow(dn,cmap='jet',clim=(-0.2, 0.2),aspect=0.2)
fig.add_subplot(3,2,5)
plt.imshow(d2,cmap='jet',clim=(-0.2, 0.2),aspect=0.2)
plt.title('Local orthogonalization');
fig.add_subplot(3,2,6)
plt.imshow(noi2,cmap='jet',clim=(-0.2, 0.2),aspect=0.2)
plt.title('Local orthogonalization');
plt.show()

fig = plt.figure(figsize=(5, 6))
fig.add_subplot(2, 1, 1)
plt.imshow(simi1,cmap='jet',clim=(0,1),aspect=0.2)
plt.title('Local similarity: Initial denoising');
fig.add_subplot(2, 1, 2)
plt.imshow(simi2,cmap='jet',clim=(0,1),aspect=0.2)
plt.title('Local similarity: Local orthogonalization');
plt.show()












test_localortho3d.py
#  DEMO script (python version) for calculating local orthogonalization and its application in random noise attenuation 
#  
#  Copyright (C) 2022 Yangkang Chen
#  
#  This program is free software: you can redistribute it and/or modify
#  it under the terms of the GNU General Public License as published
#  by the Free Software Foundation, either version 3 of the License, or
#  any later version.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
#  GNU General Public License for more details: http://www.gnu.org/licenses/
#  
#  Reference:   1. Random noise attenuation using local signal-and-noise orthogonalization
#               Chen and Fomel, 2015, Geophysics
#               2. Ground-Roll Noise Attenuation Using a Simple and Effective Approach Based on 
#               Local Band-Limited Orthogonalization, Chen et al., 2015, IEEE Geoscience and Remote Sensing Letters
#               3. Iterative deblending with multiple constraints based on shaping regularization,
#               Chen, 2015, IEEE Geoscience and Remote Sensing Letters
#               4. Orthogonalized morphological reconstruction for weak signal detection in micro-seismic monitoring:
#               Methodology, Huang et al., 2018, GJI
#               5. Surface-related multiple leakage extraction using local primary-and-multiple 
#               orthogonalization, Zhang et al., 2020, Geophysics
#               6. Non-stationary local signal-and-noise orthogonalization, Chen et al.,
#               2021, Geophysics
#               7. Local primary-and-multiple orthogonalization for leaked internal multiple crosstalk estimation and attenuation on full-wavefield migrated images
#               Zhang, et al., 2021, Geophysics

## generate synthetic data
#This synthetic data was used in Huang et al., 2016, Damped multichannel singular spectrum analysis for 3D random noise attenuation, Geophysics, 81, V261-V270.
import numpy as np
import matplotlib.pyplot as plt

## generate the synthetic data
a1=np.zeros([300,20])
[n,m]=a1.shape
a3=np.zeros([300,20])
a4=np.zeros([300,20])

k=-1;
a=0.1;
b=1;
pi=np.pi

ts=np.arange(-0.055,0.055+0.002,0.002)
b1=np.zeros([len(ts)])
b2=np.zeros([len(ts)])
b3=np.zeros([len(ts)])
b4=np.zeros([len(ts)])

for t in ts:
    k=k+1;
    b1[k]=(1-2*(pi*30*t)*(pi*30*t))*np.exp(-(pi*30*t)*(pi*30*t));
    b2[k]=(1-2*(pi*40*t)*(pi*40*t))*np.exp(-(pi*40*t)*(pi*40*t));
    b3[k]=(1-2*(pi*40*t)*(pi*40*t))*np.exp(-(pi*40*t)*(pi*40*t));
    b4[k]=(1-2*(pi*30*t)*(pi*30*t))*np.exp(-(pi*30*t)*(pi*30*t));

t1=np.zeros([m],dtype='int')
t3=np.zeros([m],dtype='int')
t4=np.zeros([m],dtype='int')
for i in range(m):
  t1[i]=np.round(140);
  t3[i]=np.round(-6*i+180);
  t4[i]=np.round(6*i+10);
  a1[t1[i]:t1[i]+k+1,i]=b1; 
  a3[t3[i]:t3[i]+k+1,i]=b1; 
  a4[t4[i]:t4[i]+k+1,i]=b1; 

temp=a1[0:300,:]+a3[0:300,:]+a4[0:300,:];

shot=np.zeros([300,20,20])
for j in range(20):
    a4=np.zeros([300,20]);
    for i in range(m):
    	t4[i]=np.round(6*i+10+3*j); 
    	a4[t4[i]:t4[i]+k+1,i]=b1;
  
    	t1[i]=np.round(140-2*j);
    	a1[t1[i]:t1[i]+k+1,i]=b1;

    shot[:,:,j]=a1[0:300,:]+a3[0:300,:]+a4[0:300,:];

d0=shot

## add noise
[n1,n2,n3]=d0.shape
np.random.seed(201415)
n=0.1*np.random.randn(n1,n2,n3);
dn=d0+n;
print(np.std(dn))

from fxydmssa import fxydmssa
d1=fxydmssa(dn,0,120,0.004,3,1);	#DMSSA (when damping factor =1, there are heavy damages)
# d1=d0*0.9
noi1=dn-d1;

from str_snr import str_snr
print('SNR of initial denoising is %g'%str_snr(d0,d1,2));


from localortho import localortho

## prepare paramters for ortho
rect=[10,10,10];
eps=0;
niter=20;
verb=1;

## calculate local orthogonalization
[d2,noi2,low]=localortho(d1,noi1,rect,niter,eps,verb);

## compare with matlab
import scipy
from scipy import io
datas = {"d0":d0,"dn": dn, "d1": d1, "noi1": noi1, "d2":d2, "noi2":noi2}
scipy.io.savemat("datas.mat", datas)


## calculate local similarity
from localsimi import localsimi
simi1=localsimi(d1,noi1,[5,5,5],niter,eps,verb);
simi2=localsimi(d2,noi2,[5,5,5],niter,eps,verb);

## plotting
fig = plt.figure(figsize=(5, 6))
fig.add_subplot(3, 2, 1)
plt.imshow(dn.transpose(0,2,1).reshape(n1,n2*n3),cmap='jet',clim=(-0.2, 0.2))
fig.add_subplot(3, 2, 3)
plt.imshow(d1.reshape(n1,n2*n3,order='F'),cmap='jet',clim=(-0.2, 0.2))
fig.add_subplot(3, 2, 4)
plt.imshow(noi1.transpose(0,2,1).reshape(n1,n2*n3),cmap='jet',clim=(-0.2, 0.2))
fig.add_subplot(3, 2, 5)
plt.imshow(d2.reshape(n1,n2*n3,order='F'),cmap='jet',clim=(-0.2, 0.2))
fig.add_subplot(3, 2, 6)
plt.imshow(noi2.transpose(0,2,1).reshape(n1,n2*n3),cmap='jet',clim=(-0.2, 0.2))
plt.show()

fig = plt.figure(figsize=(5, 6))
fig.add_subplot(2, 1, 1)
plt.imshow(simi1.reshape(n1,n2*n3,order='F'),cmap='jet',clim=(0,1))
plt.title('Local similarity: Initial denoising');
fig.add_subplot(2, 1, 2)
plt.imshow(simi2.reshape(n1,n2*n3,order='F'),cmap='jet',clim=(0,1))
plt.title('Local similarity: Local orthogonalization');
plt.show()

from str_snr import str_snr
print('SNR of initial denoising is %g'%str_snr(d0,d1,2));
print('SNR of local orthogonalization is %g'%str_snr(d0,d2,2));



test_verify_fxydmssa.m
clc;clear;close all;
%% 3D
load('datas.mat');

d11=fxydmssa(dn,0,120,0.004,3,3,0);
norm(d1(:)-d11(:))

figure; yc_imagesc([dn(:,:),d1(:,:),dn(:,:)-d1(:,:)]);
figure; yc_imagesc([d1(:,:),d11(:,:),d1(:,:)-d11(:,:)],90);

yc_snr(d0,dn,2)
yc_snr(d0,d1,2)
yc_snr(d0,d11,2)

clc;clear;close all;
%% 2D
load('datas2d.mat');
d11=fxydmssa(dn,0,120,0.004,3,3,0);
norm(d1(:)-d11(:))

figure; yc_imagesc([dn(:,:),d1(:,:),dn(:,:)-d1(:,:)]);
figure; yc_imagesc([d1(:,:),d11(:,:),d1(:,:)-d11(:,:)],90);

yc_snr(d0,dn,2)
yc_snr(d0,d1,2)
yc_snr(d0,d11,2)

str_snr.py
def str_snr(g,f,mode=1):
	#str_snr: calculate the SNR
	#This is a subroutine from the seistr package (https://github.com/chenyk1990/pyseistr)
	#
	#The definition of SNR can be found in 
	#Chen and Fomel, 2015, Random noise attenuation using local
	#signal-and-noise orthogonalization, Geophysics.
	#
	#Author: Yangkang Chen
	#g: ground truth image
	#f: noisy/restored image
	#mode:1->2D SNR, 2->3D SNR

	import numpy as np

	if g.ndim==2:
		g=np.expand_dims(g, axis=2)

	if f.ndim==2:
		f=np.expand_dims(f, axis=2)
		
	g = np.double(g); #in case of data format is unit8,12,16
	f = np.double(f);

	if f.size != g.size:
		print('Dimesion of two images don''t match!');

	if mode ==1:
		s = g.shape[2];
		if s==1: #single channel	
			psnr = 20.*np.log10(np.linalg.norm(g[:,:,0],'fro')/np.linalg.norm(g[:,:,0]-f[:,:,0],'fro'));   
		else: #multi-channel
			psnr = np.zeros(s);
		for i in range(0,s):
			psnr[i] = 20.*np.log10(np.linalg.norm(g[:,:,i],'fro')/np.linalg.norm(g[:,:,i]-f[:,:,i],'fro'));

	else:
		[n1,n2,n3]=g.shape;
		psnr = 20.*np.log10(np.linalg.norm(g.reshape(n1,n2*n3,order='F'),'fro')/np.linalg.norm(g.reshape(n1,n2*n3,order='F')-f.reshape(n1,n2*n3,order='F'),'fro'));   

	return psnr

localortho.py
def localortho(signal,noise,rect,niter=50,eps=0.0,verb=1):
	#LOCAORTHO: Noise attenuation using local signal-and-noise
	#orthogonalization and output the local orthogonalization weight (LOW)
	#
	#IN   signal:    initial signal
	#     noise:     initial noise
	#     rect:   3-D vector denoting smooth radius
	#     niter:  number of CG iterations
	#     eps:    regularization parameter, default 0.0
	#     verb:   verbosity flag (default: 0)
	#
	#OUT  signal2: orthogonalized signal
	#     noise2:  orthogonalized noise
	#     low:   local orthogonalization weight
	#
	#Copyright (C) 2016 Yangkang Chen
	#Ported to Python in 2022 by Yangkang Chen 
	#
	#This program is free software: you can redistribute it and/or modify
	#it under the terms of the GNU General Public License as published
	#by the Free Software Foundation, either version 3 of the License, or
	#any later version.
	#
	#This program is distributed in the hope that it will be useful,
	#but WITHOUT ANY WARRANTY; without even the implied warranty of
	#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	#GNU General Public License for more details: http://www.gnu.org/licenses/
	#
	#Reference:   1. Random noise attenuation using local signal-and-noise orthogonalization
	#             Chen and Fomel, 2015, Geophysics
	#             2. Ground-Roll Noise Attenuation Using a Simple and Effective Approach Based on 
	#             Local Band-Limited Orthogonalization, Chen et al., 2015, IEEE Geoscience and Remote Sensing Letters
	#             3. Iterative deblending with multiple constraints based on shaping regularization,
	#             Chen, 2015, IEEE Geoscience and Remote Sensing Letters
	#             4. Orthogonalized morphological reconstruction for weak signal detection in micro-seismic monitoring:
	#             Methodology, Huang et al., 2018, GJI
	#             5. Surface-related multiple leakage extraction using local primary-and-multiple 
	#             orthogonalization, Zhang et al., 2020, Geophysics
	#             6. Non-stationary local signal-and-noise orthogonalization, Chen et al.,
	#             2020, Geophysics
	#             7. Local primary-and-multiple orthogonalization for leaked internal multiple crosstalk estimation and attenuation on full-wavefield migrated images
	#             Zhang, et al., 2020, Geophysics
	#Example:
	#test_localortho.m

	import numpy as np
	if signal.ndim==2:	#for 2D problems
		signal=np.expand_dims(signal, axis=2)
	[n1,n2,n3]=signal.shape
	
	nd=n1*n2*n3;
	ndat=[n1,n2,n3];
	
	eps_dv=eps;
	eps_cg=0.1; 
	tol_cg=0.000001;
	from pyseistr import str_divne
	ratio = str_divne(noise, signal, niter, rect, ndat, eps_dv, eps_cg, tol_cg,verb);
	
	signal2=signal+ratio*signal;
	noise2=noise-ratio*signal;
	low=ratio;

	return signal2,noise2,low

pyortho/drr3d.py
import numpy as np
import scipy
from scipy import linalg
def drr3d(D, flow=1, fhigh=124, dt=0.004, N=1, K=1, verb=0):
	#DRR3D: 3D rank-reduction method for denoising (also known as FXYDMSSA)
	#
	#IN   D:   	 intput 3D data (ndarray)
	#     flow:   processing frequency range (lower)
	#     fhigh:  processing frequency range (higher)
	#     dt:     temporal sampling interval
	#     N:      number of singular value to be preserved
	#     K:     damping factor
	#     verb:   verbosity flag (default: 0)
	#
	#OUT  D1:  	output data
	#
	#Copyright (C) 2013 The University of Texas at Austin
	#Copyright (C) 2013 Yangkang Chen
	#Modified 2015 by Yangkang Chen
	#Ported to Python in 2022 by Yangkang Chen (Verified to be correct, the same as Matlab version)
	#
	#This program is free software: you can redistribute it and/or modify
	#it under the terms of the GNU General Public License as published
	#by the Free Software Foundation, either version 3 of the License, or
	#any later version.
	#
	#This program is distributed in the hope that it will be useful,
	#but WITHOUT ANY WARRANTY; without even the implied warranty of
	#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	#GNU General Public License for more details: http://www.gnu.org/licenses/
	#
	#References:   
	#
	#[1] Chen, Y., W. Huang, D. Zhang, W. Chen, 2016, An open-source matlab code package for improved rank-reduction 3D seismic data denoising and reconstruction, Computers & Geosciences, 95, 59-66.
	#[2] Chen, Y., D. Zhang, Z. Jin, X. Chen, S. Zu, W. Huang, and S. Gan, 2016, Simultaneous denoising and reconstruction of 5D seismic data via damped rank-reduction method, Geophysical Journal International, 206, 1695-1717.
	#[3] Huang, W., R. Wang, Y. Chen, H. Li, and S. Gan, 2016, Damped multichannel singular spectrum analysis for 3D random noise attenuation, Geophysics, 81, V261-V270.
	#[4] Chen et al., 2017, Preserving the discontinuities in least-squares reverse time migration of simultaneous-source data, Geophysics, 82, S185-S196.
	#[5] Chen et al., 2019, Obtaining free USArray data by multi-dimensional seismic reconstruction, Nature Communications, 10:4434.
	#
	# DEMO
	# demos/test_pyortho_localortho2d.py
	# demos/test_pyortho_localortho3d.py
	
	print('flow=',flow,'fhigh=',fhigh,'dt=',dt,'N=',N,'K=',K,'verb=',verb)

	if D.ndim==2:	#for 2D problems
		D=np.expand_dims(D, axis=2)

	[nt,nx,ny]=D.shape
	D1=np.zeros([nt,nx,ny])
	
	nf=2^nextpow2(nt);
	nf=int(nf)
	
	#Transform into F-X domain
	DATA_FX=np.fft.fft(D,nf,0);
	DATA_FX0=np.zeros([nf,nx,ny],dtype=np.complex_);

	#First and last nts of the DFT.
	ilow  = np.floor(flow*dt*nf)+1;
	if ilow<1:
		ilow=1
	
	ihigh = np.floor(fhigh*dt*nf)+1;
	if ihigh > np.floor(nf/2)+1:
		ihigh=np.floor(nf/2)+1
	
	ilow=int(ilow)
	ihigh=int(ihigh)
	
	lx=int(np.floor(nx/2)+1);
	lxx=nx-lx+1;
	ly=int(np.floor(ny/2)+1);
	lyy=ny-ly+1;
	M=np.zeros([lx*ly,lxx*lyy]);
	
	#main loop
	for k in range(ilow,ihigh+1):
		
		M=P_H(DATA_FX[k-1,:,:],lx,ly); 
		M=P_RD(M,N,K);
		DATA_FX0[k-1,:,:]=P_A(M,nx,ny,lx,ly);
		
		if np.mod(k,5)==0 and verb==1 : 
			print('F %d is done!\n\n'%k);
	
	for k in range(int(nf/2)+2,nf+1):
		DATA_FX0[k-1,:,:] = np.conj(DATA_FX0[nf-k+1,:,:]);

	#Back to TX (the output)
	D1=np.real(np.fft.ifft(DATA_FX0,nf,0));
	D1=D1[0:nt,:,:];
	
	if ny==1:	#for 2D problems
		D1=np.squeeze(D1)
	
	return D1
	
def nextpow2(N):
    """ Function for finding the next power of 2 """
    n = 1
    while n < N: n *= 2
    return n
    

def P_H(din,lx,ly):
	""" forming block Hankel matrix """
	[nx,ny]=din.shape;
	lxx=nx-lx+1;
	lyy=ny-ly+1;
	dout=np.zeros([lx*ly,lxx*lyy],dtype=np.complex_)
	
	for j in range(1,ny+1):
		r=scipy.linalg.hankel(din[0:lx,j-1],din[lx-1:nx,j-1]);
		if j<ly:
			for id in range(1,j+1):
				dout[(j-1)*lx-(id-1)*lx:j*lx-(id-1)*lx,(id-1)*lxx:lxx+(id-1)*lxx] = r;
		else:
			for id in range(1,ny-j+2):
				dout[(ly-1)*lx-(id-1)*lx:ly*lx-(id-1)*lx,(j-ly)*lxx+(id-1)*lxx:(j-ly+1)*lxx+(id-1)*lxx]=r;	
	return dout


def P_RD(din,N,K):
	"""Rank reduction on the block Hankel matrix"""
	[U,D,V]=scipy.linalg.svd(din)
	for j in range(1,N+1):
		D[j-1]=D[j-1]*(1-np.power(D[N],K)/np.power(D[j-1],K))
	dout=np.mat(U[:,0:N])*np.mat(np.diag(D[0:N]))*np.mat(V[0:N,:]);

	return dout
	
def P_A(din,nx,ny,lx,ly):
	""" Averaging the block Hankel matrix to output the result """ 
	lxx=nx-lx+1;
	lyy=ny-ly+1;

	dout=np.zeros([nx,ny],dtype=np.complex_);
	
	for j in range(1,ny+1):
		if j<ly:
			for id in range(1,j+1):
				dout[:,j-1] =dout[:,j-1]+ ave_antid(din[(j-1)*lx-(id-1)*lx:j*lx-(id-1)*lx,(id-1)*lxx:lxx+(id-1)*lxx])/j;
		else:
			for id in range(1,ny-j+2):
				dout[:,j-1] =dout[:,j-1]+ ave_antid(din[(ly-1)*lx-(id-1)*lx:ly*lx-(id-1)*lx,(j-ly)*lxx+(id-1)*lxx:(j-ly+1)*lxx+(id-1)*lxx])/(ny-j+1);
	
	return dout
    
def ave_antid(din):
	""" averaging along antidiagonals """
	[n1,n2]=din.shape;
	nout=n1+n2-1;
	dout=np.zeros(nout,dtype=np.complex_);

	for i in range(1,nout+1):
		if i<n1:
			for id in range(1,i+1):
				dout[i-1]=dout[i-1] + din[i-(id-1)-1,(id-1)]/i; 
		else:
			for id in range(1,nout+2-i):
				dout[i-1]=dout[i-1] + din[n1-(id-1)-1,(i-n1)+(id-1)]/(nout+1-i); 
	return dout




    


fxydmssa.py
def fxydmssa(D, flow=1, fhigh=124, dt=0.004, N=1, K=1, verb=0):
	#FXYDMSSA: F-XY domain damped multichannel singular spectrum analysis (DMSSA)
	#
	#IN   D:   	 intput 3D data (ndarray)
	#     flow:   processing frequency range (lower)
	#     fhigh:  processing frequency range (higher)
	#     dt:     temporal sampling interval
	#     N:      number of singular value to be preserved
	#     K:     damping factor
	#     verb:   verbosity flag (default: 0)
	#
	#OUT  D1:  	output data
	#
	#Copyright (C) 2013 The University of Texas at Austin
	#Copyright (C) 2013 Yangkang Chen
	#Modified 2015 by Yangkang Chen
	#Ported to Python in 2022 by Yangkang Chen (Verified to be correct, the same as Matlab version)
	#
	#This program is free software: you can redistribute it and/or modify
	#it under the terms of the GNU General Public License as published
	#by the Free Software Foundation, either version 3 of the License, or
	#any later version.
	#
	#This program is distributed in the hope that it will be useful,
	#but WITHOUT ANY WARRANTY; without even the implied warranty of
	#MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
	#GNU General Public License for more details: http://www.gnu.org/licenses/
	#
	#References:   
	#
	#[1] Chen, Y., W. Huang, D. Zhang, W. Chen, 2016, An open-source matlab code package for improved rank-reduction 3D seismic data denoising and reconstruction, Computers & Geosciences, 95, 59-66.
	#[2] Chen, Y., D. Zhang, Z. Jin, X. Chen, S. Zu, W. Huang, and S. Gan, 2016, Simultaneous denoising and reconstruction of 5D seismic data via damped rank-reduction method, Geophysical Journal International, 206, 1695-1717.
	#[3] Huang, W., R. Wang, Y. Chen, H. Li, and S. Gan, 2016, Damped multichannel singular spectrum analysis for 3D random noise attenuation, Geophysics, 81, V261-V270.
	#[4] Chen et al., 2017, Preserving the discontinuities in least-squares reverse time migration of simultaneous-source data, Geophysics, 82, S185-S196.
	#[5] Chen et al., 2019, Obtaining free USArray data by multi-dimensional seismic reconstruction, Nature Communications, 10:4434.

	print('flow=',flow,'fhigh=',fhigh,'dt=',dt,'N=',N,'K=',K,'verb=',verb)

	import numpy as np
	if D.ndim==2:	#for 2D problems
		D=np.expand_dims(D, axis=2)

	[nt,nx,ny]=D.shape
	D1=np.zeros([nt,nx,ny])
	
	nf=2^nextpow2(nt);
	nf=int(nf)
	
	#Transform into F-X domain
	DATA_FX=np.fft.fft(D,nf,0);
	DATA_FX0=np.zeros([nf,nx,ny],dtype=np.complex_);

	#First and last nts of the DFT.
	ilow  = np.floor(flow*dt*nf)+1;
	if ilow<1:
		ilow=1
	
	ihigh = np.floor(fhigh*dt*nf)+1;
	if ihigh > np.floor(nf/2)+1:
		ihigh=np.floor(nf/2)+1
	
	ilow=int(ilow)
	ihigh=int(ihigh)
	
	lx=int(np.floor(nx/2)+1);
	lxx=nx-lx+1;
	ly=int(np.floor(ny/2)+1);
	lyy=ny-ly+1;
	M=np.zeros([lx*ly,lxx*lyy]);
	
	#main loop
	for k in range(ilow,ihigh+1):
		
		M=P_H(DATA_FX[k-1,:,:],lx,ly); 
		M=P_RD(M,N,K);
		DATA_FX0[k-1,:,:]=P_A(M,nx,ny,lx,ly);
		
		if np.mod(k,5)==0 and verb==1 : 
			print('F %d is done!\n\n'%k);
	
	for k in range(int(nf/2)+2,nf+1):
		DATA_FX0[k-1,:,:] = np.conj(DATA_FX0[nf-k+1,:,:]);

	#Back to TX (the output)
	D1=np.real(np.fft.ifft(DATA_FX0,nf,0));
	D1=D1[0:nt,:,:];
	
	if ny==1:	#for 2D problems
		D1=np.squeeze(D1)
	
	return D1
	
def nextpow2(N):
    """ Function for finding the next power of 2 """
    n = 1
    while n < N: n *= 2
    return n
    

def P_H(din,lx,ly):
	""" forming block Hankel matrix """
	import numpy as np
	import scipy
	from scipy import linalg
	[nx,ny]=din.shape;
	lxx=nx-lx+1;
	lyy=ny-ly+1;
	dout=np.zeros([lx*ly,lxx*lyy],dtype=np.complex_)
	
	for j in range(1,ny+1):
		r=scipy.linalg.hankel(din[0:lx,j-1],din[lx-1:nx,j-1]);
		if j<ly:
			for id in range(1,j+1):
				dout[(j-1)*lx-(id-1)*lx:j*lx-(id-1)*lx,(id-1)*lxx:lxx+(id-1)*lxx] = r;
		else:
			for id in range(1,ny-j+2):
				dout[(ly-1)*lx-(id-1)*lx:ly*lx-(id-1)*lx,(j-ly)*lxx+(id-1)*lxx:(j-ly+1)*lxx+(id-1)*lxx]=r;	
	return dout


def P_RD(din,N,K):
	"""Rank reduction on the block Hankel matrix"""
	import scipy
	from scipy import linalg
	import numpy as np
	[U,D,V]=scipy.linalg.svd(din)
	for j in range(1,N+1):
		D[j-1]=D[j-1]*(1-np.power(D[N],K)/np.power(D[j-1],K))
	dout=np.mat(U[:,0:N])*np.mat(np.diag(D[0:N]))*np.mat(V[0:N,:]);

	return dout
	
def P_A(din,nx,ny,lx,ly):
	""" Averaging the block Hankel matrix to output the result """ 
	import numpy as np
	lxx=nx-lx+1;
	lyy=ny-ly+1;

	dout=np.zeros([nx,ny],dtype=np.complex_);
	
	for j in range(1,ny+1):
		if j<ly:
			for id in range(1,j+1):
				dout[:,j-1] =dout[:,j-1]+ ave_antid(din[(j-1)*lx-(id-1)*lx:j*lx-(id-1)*lx,(id-1)*lxx:lxx+(id-1)*lxx])/j;
		else:
			for id in range(1,ny-j+2):
				dout[:,j-1] =dout[:,j-1]+ ave_antid(din[(ly-1)*lx-(id-1)*lx:ly*lx-(id-1)*lx,(j-ly)*lxx+(id-1)*lxx:(j-ly+1)*lxx+(id-1)*lxx])/(ny-j+1);
	
	return dout
    
def ave_antid(din):
	""" averaging along antidiagonals """
	import numpy as np
	[n1,n2]=din.shape;
	nout=n1+n2-1;
	dout=np.zeros(nout,dtype=np.complex_);

	for i in range(1,nout+1):
		if i<n1:
			for id in range(1,i+1):
				dout[i-1]=dout[i-1] + din[i-(id-1)-1,(id-1)]/i; 
		else:
			for id in range(1,nout+2-i):
				dout[i-1]=dout[i-1] + din[n1-(id-1)-1,(i-n1)+(id-1)]/(nout+1-i); 
	return dout




    
